### PR TITLE
Sharding fails in TF when absolute scope was modified if `.` in layer name

### LINKED ISSUE #18776: TF: Can't create sharded XGLM model
### System Info

- `transformers` version: 4.22.0.dev0
- Platform: Linux-5.15.0-33-generic-x86_64-with-glibc2.35
- Python version: 3.8.13
- Huggingface_hub version: 0.9.0
- PyTorch version (GPU?): 1.12.0+cu116 (True)
- Tensorflow version (GPU?): 2.9.1 (True)
- Flax version (CPU?/GPU?/TPU?): 0.5.0 (gpu)
- Jax version: 0.3.5
- JaxLib version: 0.3.5
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Running this CLI command

```
CUDA_VISIBLE_DEVICES="" TOKENIZERS_PARALLELISM=false NVIDIA_TF32_OVERRIDE=0 transformers-cli pt-to-tf --model-name facebook/xglm-2.9B --new-weights --max-error 3e-3
```

Gets you the following exception (in the sharding code)

```
Traceback (most recent call last):
  File "/home/joao/hf/bin/transformers-cli", line 8, in <module>
    sys.exit(main())
  File "/home/joao/transformers/src/transformers/commands/transformers_cli.py", line 55, in main
    service.run()
  File "/home/joao/transformers/src/transformers/commands/pt_to_tf.py", line 309, in run
    tf_from_pt_model.save_pretrained(self._local_dir)
  File "/home/joao/transformers/src/transformers/modeling_tf_utils.py", line 2020, in save_pretrained
    param_dset = shard_file.create_dataset(
  File "/home/joao/hf/lib/python3.8/site-packages/h5py/_hl/group.py", line 161, in create_dataset
    dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
  File "/home/joao/hf/lib/python3.8/site-packages/h5py/_hl/dataset.py", line 156, in make_new_dset
    dset_id = h5d.create(parent.id, name, tid, sid, dcpl=dcpl, dapl=dapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5d.pyx", line 84, in h5py.h5d.create
TypeError: expected bytes, str found
```

### Expected behavior

Successful sharding :D

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?

Fixes #18776, by taking care of the particular case of absolute scope modifications

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Should probably add a test for that

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -2151,17 +2158,22 @@ def save_pretrained(
...
-                            "/".join(layer.name.split("/")[1:]), layer.numpy().shape, dtype=layer.numpy().dtype
+                            layer_name, layer.numpy().shape, dtype=layer.numpy().dtype
                         )
                         param_dset[:] = layer.numpy()
+                        layers.append(layer_name.encode("utf8"))
+                    save_attributes_to_hdf5_group(
+                        shard_file,
+                        "layer_names",
+                        layers,
+                    )
```
Comment (sgugger) [inline src/transformers/modeling_tf_utils.py:2176]: This fits in one line now.

**Review by sgugger — COMMENTED** [review]
Yes, there should be a test of the current failing behavior pretty please :pray: 
Also, @gante and/or @Rocketknight1 I'd appreciate if you could have a look here.

**Review by gante — APPROVED** [review]
It looks good to me, but can you confirm (or better, add a test!) that it works in complex models like RAG? 🙏 

(RAG is a model where one of the components, the generator, might have these shared weights)

**Comment by ArthurZucker** [comment]
On it! 🤗

**Comment by ArthurZucker** [comment]
Just added a test, RAG works, I think I'll try to test most of our models just to be sure that there is no other strange pattern.

**Review by Rocketknight1 — APPROVED** [review]
LGTM! I think this is the last issue left with my TF port of ESM too, so it's greatly appreciated!

**Comment by ArthurZucker** [comment]
Tested with `"openai/clip-vit-large-patch14",  "xlm-roberta-base"` in addition, works nicely.

**Review by sgugger — APPROVED** [review]

**Pull request merged**